### PR TITLE
Gen4: improve expand star

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This pull request improves the expand star logic. Star expressions are now supported in the select list.

While improving the expand star logic, @systay and I improved the analyzer. The analyzer is now binding columns to expressions in a better way, mainly a new `dependencies` interface allows us to tell three things: if a column has `certain`, `uncertain` dependencies or none (`nothing`), this gives us more power when trying to detect column ambiguity. Additionally, the derived tables are using a different implementation `TableInfo` which is `derivedTable` instead of `vTableInfo`.

### New supported queries

- select star against a derived table selecting from an unsharded tabled: `select u.* from (select * from unsharded) u`
- select star against a sharded table: `select user.* from user`
- select star inside a derived table against information schema table (non authoritative): `select x.table_name from (select a.* from information_schema.key_column_usage a) x`
- join with a predicate that comes from a derived table that does a select star against an information schema table: `select x.table_name from (select a.* from information_schema.key_column_usage a) x join user on x.id = user.id`

## Related Issue(s)

- #7280 

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/derivedtree.go [hunk 1]
```diff
@@ -45,35 +48,59 @@ func (d *derivedTree) clone() queryTree {
...
+			hasStar = true
 		}
 	}
-	return 0, vterrors.NewErrorf(vtrpcpb.Code_NOT_FOUND, vterrors.BadFieldError, "Unknown column '%s' in 'field list'", name.Name.String())
+
+	// we have found a star but no matching *sqlparser.AliasedExpr, thus we return -1 with no error.
+	if hasStar {
+		return -1, nil
+	}
+	return -1, vterrors.NewErrorf(vtrpcpb.Code_NOT_FOUND, vterrors.BadFieldError, "Unknown column '%s' in 'field list'", name.Name.String())
```
Comment (systay) [inline go/vt/vtgate/planbuilder/derivedtree.go:105]: nit: should we not return `0` here? not at all important, just looks wrong

File: go/vt/vtgate/planbuilder/testdata/systemtables_cases.txt [hunk 2]
```diff
@@ -705,6 +722,44 @@ Gen4 plan same as above
 }
 Gen4 plan same as above
 
+# subquery of information_schema with itself and star expression in outer select
+"select a.*, u.id from information_schema.a a, user u where a.id in (select * from information_schema.b)"
+"unsupported: '*' expression in cross-shard query"
+{
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/systemtables_cases.txt:728]: I think this is wrong. The join primitive can't handle `*` like this

File: go/vt/sqlparser/ast_funcs.go [hunk 3]
```diff
@@ -686,6 +686,10 @@ func (node *TableIdent) UnmarshalJSON(b []byte) error {
 func containEscapableChars(s string, at AtCount) bool {
 	isDbSystemVariable := at != NoAt
 
+	// stars (i.e "*") are not escapable
```
Comment (systay) [inline go/vt/sqlparser/ast_funcs.go:689]: this has got to go

**Review by systay — APPROVED** [review]
Nice stuff!
Solving the queries, and added lots of helpful comments as well :star:

File: go/vt/vtgate/planbuilder/abstract/queryprojection.go [hunk 4]
```diff
@@ -61,33 +62,42 @@ type (
...
+}
+
+// GetAliasedExpr returns the SelectExpr as a *sqlparser.AliasedExpr if its type allows it,
+// otherwise an error is returned.
+func (s SelectExpr) GetAliasedExpr() (*sqlparser.AliasedExpr, error) {
+	switch expr := s.Col.(type) {
+	case *sqlparser.AliasedExpr:
+		return expr, nil
+	case *sqlparser.StarExpr:
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: '*' expression in cross-shard query")
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/abstract/queryprojection.go:85]: This seems wrong, we should probably return a global error and catch it from the outside and decide whether we want to return an `expression in cross-shard query` error, like we did for multiple tables predicates

File: go/vt/vtgate/semantics/analyzer_test.go [hunk 5]
```diff
@@ -37,14 +34,35 @@ const (
...
+
+   select x.id, z.col, x.col from (select tbl1.a+tbl2.b as id, 1 as col from tvl1, tbl2) as x, z
+
+   x.id : direct dependencies: x, recursive dependencies: tbl1, tbl2
+   z.col: direct dependencies: z, recursive dependencies: z
+   x.col: direct dependencies: x, recursive dependencies: 0
+
+
+
+*/
```
Comment (frouioui) [inline go/vt/vtgate/semantics/analyzer_test.go:64]: Let's add this to a new issue to keep track of this test matrix and remove the comment

File: go/vt/vtgate/semantics/dependencies.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+type (
+	dependencies interface {
+		Empty() bool
+		Get() (dependency, error)
+		Merge(dependencies) (dependencies, error)
+	}
+	dependency struct {
+		direct    TableSet
+		recursive TableSet
```
Comment (frouioui) [inline go/vt/vtgate/semantics/dependencies.go:33]: Probably in a separate PR: we need to make sure we change the names of our dependencies from `base` to `recursive`, and `real|deps` to `direct` throughout the whole codebase

Comment (frouioui) [reply go/vt/vtgate/semantics/dependencies.go:33]: Resolved via eb1e967

**Review by harshit-gangal — APPROVED** [review]
overall looks good.

File: go/vt/vtgate/planbuilder/testdata/postprocess_cases.txt [hunk 7]
```diff
@@ -1171,83 +1172,6 @@ Gen4 plan same as above
 }
 Gen4 plan same as above
 
-# Order by, '*' expression in a subquery
-"select u.id, e.id from user u join user_extra e where u.col = e.col and u.col in (select * from user where user.id = u.id order by col)"
```
[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/vtgate/planbuilder/testdata/postprocess_cases.txt]: no, it was not on purpose

**Review by harshit-gangal — COMMENTED** [review]
changes made to go/vt/topo/k8stopo packages seems unrelated to this PR.

**Comment by frouioui** [comment]
The Git hook prompted me to run the `goimports` command on those files after I merged main into this branch.

**Pull request merged**